alacritty/src/renderer/solidrect.rs

```diff
@@ -0,0 +1,200 @@
+
+            gl::BindVertexArray(vao);
+            gl::BindBuffer(gl::ELEMENT_ARRAY_BUFFER, ebo);
+
+            gl::BindBuffer(gl::ARRAY_BUFFER, vbo);
+            // Position.
+            gl::VertexAttribPointer(
+                0,
+                2,
+                gl::FLOAT,
+                gl::FALSE,
+                (size_of::<Vertex>()) as _,
+                ptr::null(),
+            );
+            gl::EnableVertexAttribArray(0);
+
+            // Color
+            gl::VertexAttribPointer(
+                1,
+                4,
+                gl::UNSIGNED_BYTE,
+                gl::TRUE,
+                (size_of::<Vertex>()) as _,
+                offset_of!(Vertex, color) as *const _,
+            );
+            gl::EnableVertexAttribArray(1);
+
+            gl::BindVertexArray(0);
+            gl::BindBuffer(gl::ARRAY_BUFFER, 0);
+        }
+
+        Self { vao, vbo, ebo, indices: Vec::new(), vertices: Vec::new(), uploaded_indices: 0 }
+    }
+
+    pub fn draw(
+        &mut self,
+        program: &RectShaderProgram,
+        size_info: &SizeInfo,
+        rects: Vec<RenderRect>,
+    ) {
+        if rects.is_empty() {
+            return;
+        }
+
+        // Prepare common state
+        unsafe {
+            // Remove padding from viewport.
+            gl::Viewport(0, 0, size_info.width() as i32, size_info.height() as i32);
+
+            gl::Enable(gl::BLEND);
+            gl::BlendFuncSeparate(gl::SRC_ALPHA, gl::ONE_MINUS_SRC_ALPHA, gl::SRC_ALPHA, gl::ONE);
+
+            // Setup bindings. VAO will set up attribs and EBO, but not VBO.
+            gl::BindVertexArray(self.vao);
+            gl::BindBuffer(gl::ARRAY_BUFFER, self.vbo);
+
+            gl::UseProgram(program.id);
+        }
+
+        let center_x = size_info.width() / 2.;
+        let center_y = size_info.height() / 2.;
+
+        for rect in &rects {
+            self.append_rect(center_x, center_y, rect);
+        }
+
+        self.draw_accumulated();
+    }
+
+    fn append_rect(&mut self, center_x: f32, center_y: f32, rect: &RenderRect) {
+        if rect.alpha <= 0. {
+            return;
+        }
+
+        assert!(self.vertices.len() <= MAX_U16_INDICES - 4);
+
+        // Calculate rectangle position.
+        let x = (rect.x - center_x) / center_x;
+        let y = -(rect.y - center_y) / center_y;
+        let width = rect.width / center_x;
+        let height = rect.height / center_y;
+        let color = Rgba {
+            r: rect.color.r,
+            g: rect.color.g,
+            b: rect.color.b,
+            a: (rect.alpha * 255.) as u8,
+        };
+
+        self.vertices.extend_from_slice(&[
+            Vertex { x, y, color },
+            Vertex { x, y: y - height, color },
+            Vertex { x: x + width, y, color },
+            Vertex { x: x + width, y: y - height, color },
+        ]);
+
+        if self.vertices.len() == MAX_U16_INDICES {
+            self.draw_accumulated();
+        }
```

Comment:
What's the reason behind limiting this at all? Would it impact performance due to the payload being too big, or is it solely about memory usage?

Reply:
GLES2 spec only allows 16-bit indices, and I still hope to be able to support that.